05-mvc-with-rails/error-handling/readme.md

```diff
@@ -3,14 +3,388 @@
+
+```rb
+class ArtistsController < ApplicationController
+
+  def index
+    "batman" - 42
+    @artists = Artist.all
+  end
+```
+
+Go to `localhost:3000/artists` and you should get an "undefined method" error.
+
+Now put it inside a `begin/rescue` block:
+
+```rb
+class ArtistsController < ApplicationController
+
+  def index
+    begin
+      "batman" - 42
+    rescue
+    end
+    @artists = Artist.all
+  end
+```
+
+What happens when you refresh the page? It loads!
+
+The `begin` tells Ruby, "Start listening for errors." The `rescue` tells Ruby to "rescue" the app from breaking when any errors come up.
+
+**Every `begin` must have a `rescue`, and vice-versa. It must also have an `end`.**
+
+### Error messages
+
+That's all very well, but it's not very useful.
+
+#### Why is this error-suppresser not very useful?
+
+If someone's interacting with your app and does something wrong, they want to know what they did wrong. Silent errors are just as bad as app-breaking errors!
+
+Let's make an error message that actually shows up somewhere:
+
+```rb
+class ArtistsController < ApplicationController
+
+  def index
+    begin
+      "batman" - 42
+    rescue
+      @error = "You done broke it."
+    end
+    @artists = Artist.all
+  end
+```
+
+We can make errors show up in the application layout:
+
+```erb
+<body>
+  <h1>Tun.r</h1>
+  <p><%= @error %></p>
+```
+
+That at least tells us something went wrong, but doesn't tell us *what*.
+
+### StandardError
+
+We can tell Ruby to listen for a specific error type, and then when an error of that type happens, to save the information about the error to a variable:
+
+```rb
+class ArtistsController < ApplicationController
+
+  def index
+    begin
+      "batman" - 42
+    rescue StandardError => sad_panda
+      @error = sad_panda.message
+    end
+    @artists = Artist.all
+  end
+```
+
+Ruby has many different types of errors, and each one has a class. You can even make your own! However, `StandardError` is the class from which almost all of them inherit, so that's what we're using here.
+
+We're saving the `StandardError` into a variable called `sad_panda` with the hashrocket `=>` notation. This variable can be called anything.
+
+### Adding in validations
+
+Lets make an actually useful error. Change the `index` action back to what it was before:
+
+```rb
+  def index
+    @artists = Artist.all
+  end
+```
+
+Let's add a `begin/rescue` to the `create` action:
+
+```rb
+  def create
+    begin
+      @artist = Artist.create!(artist_params)
+    rescue StandardError => sad_panda
+      @error = sad_panda.message
+    end
+    redirect_to "/artists/#{@artist.id}"
+  end
+```
+
+Now, let's add a simple validation to the Artist model:
+
+```rb
+class Artist < ActiveRecord::Base
+  has_many :songs
+  validates :name, presence: true
+end
+```
+
+...and to trigger it, submit the form with nothing in the `name` field.
+
```

Comment:
definitely recommend something like this for validation-specific errors https://github.com/jshawl/ts/blob/5e1ec1257521e2b395f3d9bf44c08d698b8b2b65/app/views/tshirts/_form.html.erb#L2-L12